compiler/src/iree/compiler/DispatchCreation/CollapseDimensions.cpp

```diff
@@ -405,19 +408,133 @@ CollapseInfo::getTransformedCollapsableLoops(AffineMap map) const {
   return transformedLoops;
 }
 
-// Update `collapsableLoops` by taking the set intersection with
-// `otherCollapsable` and update the reassociation indicies accordingly.
-bool CollapseInfo::updateCollapseViaIntersect(
-    const CollapsableLoopsSet &otherCollapsable) {
-  CollapsableLoopsSet toRemove;
-  for (auto elem : collapsableLoops) {
-    if (!otherCollapsable.contains(elem)) {
-      toRemove.insert(elem);
+FailureOr<SmallVector<ReassociationIndices>>
+CollapseInfo::getTransformedReassociation(AffineMap map) const {
+  if (!map) {
+    return failure();
+  }
+
+  SmallVector<ReassociationIndices> transformedReassociation(reassociation);
+  for (const auto &[idxOne, indicies] : llvm::enumerate(reassociation)) {
+    for (auto [idxTwo, elem] : llvm::enumerate(indicies)) {
+      assert(elem < map.getNumResults() && "index has no valid mapping");
+      auto dimExpr = dyn_cast<AffineDimExpr>(map.getResult(elem));
+      if (!dimExpr) {
+        continue;
+      }
+      transformedReassociation[idxOne][idxTwo] = dimExpr.getPosition();
     }
   }
-  collapsableLoops.set_subtract(toRemove);
-  updateReassociation();
-  return toRemove.size();
+  return transformedReassociation;
+}
+
+// Update CollapseInfo to ensure that all dimensions collapsable in `this` are
+// also collapsable in `otherInfo`. This means:
+// 1. Any dimension not collapsable in `otherInfo` should not be collapsable in
+// `this`
+// 2. For any pair of dimensions in `this`, if they are collapsable in
+// `consumerInfo`, they must be collapsable into the same dimension in
+// `consumerInfo` to be collapsable into the same dimension in `this`.
+bool CollapseInfo::updateFromConsumer(OpOperand *operand,
+                                      const CollapseInfo &consumerInfo) {
+  bool didChange = false;
+  FailureOr<AffineMap> consumerToProducerMap =
+      getConsumerLoopToProducerLoopsMap(*operand);
+  if (failed(consumerToProducerMap)) {
+    this->clear();
+    return !consumerInfo.getCollapsibleLoops().empty();
+  }
+
+  FailureOr<CollapsableLoopsSet> consumerCollapsable =
+      consumerInfo.getTransformedCollapsableLoops(
+          consumerToProducerMap.value());
+  if (failed(consumerCollapsable)) {
+    didChange = collapsableLoops.size();
+    this->clear();
+    return didChange;
```

Comment:
we can simplify it a bit.

```suggestion
    this->clear();
    return !collapsableLoops.empty();
```

Reply:
I dont think this works because `clear` clears collapseableLoops

Reply:
I see, thanks! I need to clarify what the return value means in the method. If returning true means that the internal data is changed, we perhaps can make `clear` to return a boolean, then it can be simpler.